airgun/browser.py

```diff
@@ -49,59 +65,316 @@ class BrowserPlugin(DefaultPlugin):
         '''
 
 
-def browser(browser_name=None, webdriver_name=None):
-    """Creates a webdriver browser.
+class SeleniumBrowserFactory(object):
+    """
 
-    :param browser_name: one of selenium, saucelabs, docker
-    :param webdriver_name: one of firefox, chrome, edge, ie, phantomjs
+    Usage:
+        # init factory
+        factory = SeleniumBrowserFactory(test_name=test_name)
+        # get factory browser
+        selenium_browser = factory.get_browser()
+        # navigate to desired url
+        # ...
+        # perform post-init hacks (e.g. skipping certificate error screen)
+        factory.post_init()
+        # perform your test steps
+        # ...
+        # perform factory clean-up
+        factory.finalize(passed)
 
-    If any of the params are None then will be read from properties file.
     """
 
-    webdriver_name = (webdriver_name or settings.selenium.webdriver).lower()
-    browser_name = (browser_name or settings.selenium.browser).lower()
-
-    if browser_name == 'selenium':
-        if webdriver_name == 'firefox':
-            return webdriver.Firefox()
-        elif webdriver_name == 'chrome':
-            return (
-                webdriver.Chrome() if (
-                    settings.selenium.webdriver_binary is None)
-                else webdriver.Chrome(
-                    executable_path=settings.selenium.webdriver_binary)
+    def __init__(self, browser_name=None, webdriver_name=None, test_name=None):
+        self.browser = browser_name or settings.selenium.browser or 'selenium'
+        self.webdriver = (
+            webdriver_name or settings.selenium.webdriver or 'chrome')
+        self.test_name = test_name
+        self._docker = None
+        self._webdriver = None
+
+    def get_browser(self):
+        if self.browser == 'selenium':
+            return self.get_selenium_browser()
+        elif self.browser == 'saucelabs':
+            return self.get_saucelabs_browser()
+        elif self.browser == 'docker':
+            return self.get_docker_browser()
+        else:
+            raise ValueError(
+                '"{}" browser is not supported. Please use one of {}'
+                .format(self.browser, ('selenium', 'saucelabs', 'docker'))
             )
-        elif webdriver_name == 'ie':
-            return (
-                webdriver.Ie() if settings.selenium.webdriver_binary is None
-                else webdriver.Ie(
-                    executable_path=settings.selenium.webdriver_binary)
+
+    def post_init(self):
+        # Workaround 'Certificate Error' screen on Microsoft Edge
```

Comment:
I think we can say here that we can do something more than just do workaround for IE